src/lib/gadgets/bitwise.ts

```diff
@@ -0,0 +1,131 @@
+import { Provable } from '../provable.js';
+import { Field as Fp } from '../../provable/field-bigint.js';
+import { Field } from '../field.js';
+import * as Gates from '../gates.js';
+
+export { xor };
+
+function xor(a: Field, b: Field, length: number) {
+  // check that both input lengths are positive
+  assert(length > 0, `Input lengths need to be positive values.`);
+
+  // check that length does not exceed maximum field size in bits
+  assert(
+    length <= Field.sizeInBits(),
+    `Length ${length} exceeds maximum of ${Field.sizeInBits()} bits.`
+  );
+
+  // obtain pad length until the length is a multiple of 16 for n-bit length lookup table
+  let padLength = Math.ceil(length / 16) * 16;
+
+  // handle constant case
+  if (a.isConstant() && b.isConstant()) {
+    let max = 1n << BigInt(padLength);
+
+    assert(
+      a.toBigInt() < max,
+      `${a.toBigInt()} does not fit into ${padLength} bits`
+    );
+
+    assert(
+      b.toBigInt() < max,
+      `${b.toBigInt()} does not fit into ${padLength} bits`
+    );
+
+    return new Field(Fp.xor(a.toBigInt(), b.toBigInt()));
```

Comment:
I am not familiar with the types in o1js so this might be a basic question. Is converting to BigInt the desired behavior here? Since the inputs are `Field`, I wonder why this conversion needs to take place.

Reply:
This case only happens if both elements are constants (not variables that the prover needs to provide). Because of that, we can just convert them to bigint and do our computation there and convert them back. Doing so is more efficient than going through snarky and letting snarky do the calculation on constants. This is a change we did somewhat recently: Computations on constants are done on native javascript types